File: backend/app/services/energy_targets.py
"""Daily energy targets from anthropometry + goals (Mifflin–St Jeor)."""

from __future__ import annotations

from datetime import date
from typing import Any


ACTIVITY_FACTORS: dict[str, float] = {
    "sedentary": 1.2,
    "light": 1.375,
    "moderate": 1.55,
    "active": 1.725,
    "very_active": 1.9,
}

# Default calorie adjustment % by primary goal if user did not set explicit %
GOAL_DEFAULT_PCT: dict[str, float] = {
    "lose_fat": -15.0,
    "gain_muscle": 10.0,
    "maintain": 0.0,
}

MANUAL_CALORIE_TARGET_MIN = 800.0
MANUAL_CALORIE_TARGET_MAX = 10_000.0


def age_from_birth_date(birth_date: str | date | None, today: date | None = None) -> int | None:
    if birth_date is None:
        return None
    if isinstance(birth_date, str):
        try:
            birth = date.fromisoformat(birth_date[:10])
        except ValueError:
            return None
    else:
        birth = birth_date
    ref = today or date.today()
    years = ref.year - birth.year - ((ref.month, ref.day) < (birth.month, birth.day))
    if years < 10 or years > 100:
        return None
    return years


def resolve_age(anthropometry: dict[str, Any], today: date | None = None) -> int | None:
    birth = anthropometry.get("birth_date")
    from_birth = age_from_birth_date(birth, today=today)
    if from_birth is not None:
        return from_birth
    raw = anthropometry.get("age")
    try:
        age = int(raw)
    except (TypeError, ValueError):
        return None
    if 10 <= age <= 100:
        return age
    return None


def is_female_sex(sex: str | None) -> bool:
    """Normalize sex labels (en/ru) for Mifflin–St Jeor and calorie floor."""
    s = (sex or "").strip().lower().replace("ё", "е")
    if not s:
        return False
    if s in {"f", "female", "woman", "w", "ж", "жен", "женский", "female_sex"}:
        return True
    # prefixes: "жен...", "female..."
    if s.startswith("f") and "male" not in s:
        # "f", "female", "fem" — but not "male"
        if s == "f" or s.startswith("fem"):
            return True
    if s.startswith("ж"):
        return True
    return False


def is_male_sex(sex: object) -> bool:
    value = str(sex or "").strip().lower().replace("ё", "е")
    return value in {"m", "male", "man", "м", "муж", "мужской"} or value.startswith("муж")


def is_unspecified_sex(sex: object) -> bool:
    value = str(sex or "").strip().lower()
    return value in {"", "unspecified", "not_specified"} or not (
        is_female_sex(value) or is_male_sex(value)
    )


def manual_calorie_target(goals: dict[str, Any]) -> float | None:
    raw = goals.get("manual_calorie_target")
    try:
        value = float(raw)
    except (TypeError, ValueError):
        return None
    if MANUAL_CALORIE_TARGET_MIN <= value <= MANUAL_CALORIE_TARGET_MAX:
        return value
    return None


def mifflin_st_jeor_bmr(*, sex: str, weight_kg: float, height_cm: float, age: int) -> float:
    """BMR kcal/day (Mifflin–St Jeor, 1990).

    Men:   BMR = 10·W(kg) + 6.25·H(cm) − 5·A(y) + 5
    Women: BMR = 10·W(kg) + 6.25·H(cm) − 5·A(y) − 161
    """
    if is_female_sex(sex):
        return 10.0 * weight_kg + 6.25 * height_cm - 5.0 * age - 161.0
    return 10.0 * weight_kg + 6.25 * height_cm - 5.0 * age + 5.0


def resolve_adjustment_pct(goals: dict[str, Any]) -> float:
    raw = goals.get("calorie_adjustment_pct")
    if raw is not None and raw != "":
        try:
            pct = float(raw)
            return max(-40.0, min(40.0, pct))
        except (TypeError, ValueError):
            pass
    goal = str(goals.get("primary_goal") or "maintain").lower()
    return GOAL_DEFAULT_PCT.get(goal, 0.0)


def resolve_activity_level(goals: dict[str, Any], anthropometry: dict[str, Any]) -> str:
    level = str(
        goals.get("activity_level")
        or anthropometry.get("activity_level")
        or "moderate"
    ).lower()
    if level not in ACTIVITY_FACTORS:
        # map training days roughly
        try:
            days = int(goals.get("days_per_week") or 3)
        except (TypeError, ValueError):
            days = 3
        if days <= 2:
            return "light"
        if days <= 4:
            return "moderate"
        if days <= 5:
            return "active"
        return "very_active"
    return level


def macro_split_grams(*, calories: float, weight_kg: float, primary_goal: str) -> dict[str, float]:
    """Simple macro targets (g). Protein ~1.6–2.0 g/kg, fat ~25–30% kcal, rest carbs."""
    goal = (primary_goal or "maintain").lower()
    protein_per_kg = 2.0 if goal == "lose_fat" else 1.8 if goal == "gain_muscle" else 1.6
    protein_g = max(80.0, protein_per_kg * weight_kg)
    fat_ratio = 0.25 if goal == "gain_muscle" else 0.30
    fat_g = max(40.0, (calories * fat_ratio) / 9.0)
    protein_kcal = protein_g * 4.0
    fat_kcal = fat_g * 9.0
    carbs_g = max(0.0, (calories - protein_kcal - fat_kcal) / 4.0)
    return {
        "proteins_g": round(protein_g, 1),
        "fats_g": round(fat_g, 1),
        "carbs_g": round(carbs_g, 1),
    }


def compute_energy_targets(
    anthropometry: dict[str, Any] | None,
    goals: dict[str, Any] | None,
    *,
    today: date | None = None,
) -> dict[str, Any]:
    """Return BMR/TDEE/target calories + macros. incomplete=True if inputs missing."""
    anthro = anthropometry or {}
    g = goals or {}

    sex_raw = anthro.get("sex") or g.get("sex")
    if is_unspecified_sex(sex_raw):
        target = manual_calorie_target(g)
        if target is None:
            return {
                "complete": False,
                "reason": "need_sex_or_manual_target",
                "formula": None,
                "sex": None,
                "calories_target": None,
                "bmr": None,
                "tdee": None,
            }
        primary = str(g.get("primary_goal") or "maintain")
        try:
            weight = float(anthro.get("weight_kg"))
        except (TypeError, ValueError):
            weight = None
        macros = (
            macro_split_grams(calories=target, weight_kg=weight, primary_goal=primary)
            if weight is not None and weight > 0
            else None
        )
        return {
            "complete": True,
            "reason": None,
            "formula": "manual",
            "sex": None,
            "age": resolve_age(anthro, today=today),
            "birth_date": anthro.get("birth_date"),
            "weight_kg": round(weight, 2) if weight is not None else None,
            "height_cm": anthro.get("height_cm"),
            "activity_level": resolve_activity_level(g, anthro),
            "activity_factor": None,
            "bmr": None,
            "tdee": None,
            "calorie_adjustment_pct": None,
            "calories_target": round(target, 0),
            "macros": macros,
            "primary_goal": primary,
        }

    try:
        weight = float(anthro.get("weight_kg"))
        height = float(anthro.get("height_cm"))
    except (TypeError, ValueError):
        return {
            "complete": False,
            "reason": "need_weight_height",
            "calories_target": None,
            "bmr": None,
            "tdee": None,
        }

    age = resolve_age(anthro, today=today)
    if age is None:
        return {
            "complete": False,
            "reason": "need_age_or_birth_date",
            "calories_target": None,
            "bmr": None,
            "tdee": None,
            "weight_kg": weight,
            "height_cm": height,
        }

    sex = str(sex_raw)
    female = is_female_sex(sex)
    activity = resolve_activity_level(g, anthro)
    factor = ACTIVITY_FACTORS[activity]
    bmr = mifflin_st_jeor_bmr(sex=sex, weight_kg=weight, height_cm=height, age=age)
    tdee = bmr * factor
    adj_pct = resolve_adjustment_pct(g)
    target = tdee * (1.0 + adj_pct / 100.0)
    # safety floor (common clinical minimums; not part of Mifflin–St Jeor itself)
    target = max(1200.0 if female else 1500.0, target)
    primary = str(g.get("primary_goal") or "maintain")
    macros = macro_split_grams(calories=target, weight_kg=weight, primary_goal=primary)

    return {
        "complete": True,
        "reason": None,
        "formula": "mifflin_st_jeor",
        "sex": sex,
        "age": age,
        "birth_date": anthro.get("birth_date"),
        "weight_kg": round(weight, 2),
        "height_cm": round(height, 1),
        "activity_level": activity,
        "activity_factor": factor,
        "bmr": round(bmr, 0),
        "tdee": round(tdee, 0),
        "calorie_adjustment_pct": adj_pct,
        "calories_target": round(target, 0),
        "macros": macros,
        "primary_goal": primary,
    }

File: frontend/src/utils/energyTargets.ts
/**
 * Client-side Mifflin–St Jeor helpers (mirrors backend energy_targets).
 * Used for live preview in onboarding/profile before save.
 */

export type EnergyInput = {
  sex?: string | null;
  weightKg?: number | null;
  heightCm?: number | null;
  age?: number | null;
  birthDate?: string | null;
  activityLevel?: string | null;
  daysPerWeek?: number | null;
  primaryGoal?: string | null;
  calorieAdjustmentPct?: number | null;
};

const ACTIVITY: Record<string, number> = {
  sedentary: 1.2,
  light: 1.375,
  moderate: 1.55,
  active: 1.725,
  very_active: 1.9,
};

const GOAL_PCT: Record<string, number> = {
  lose_fat: -15,
  gain_muscle: 10,
  maintain: 0,
};

/**
 * Parse YYYY-MM-DD (or ISO datetime) as a local calendar date.
 * Avoid `new Date("1990-05-01")` which is UTC midnight and can shift the day
 * (and thus age/year) in timezones east of UTC.
 */
export function parseLocalDateInput(value: string | null | undefined): Date | null {
  if (!value) return null;
  const raw = String(value).trim();
  if (!raw) return null;
  const m = /^(\d{4})-(\d{2})-(\d{2})/.exec(raw);
  if (m) {
    const y = Number(m[1]);
    const mo = Number(m[2]);
    const d = Number(m[3]);
    if (!y || mo < 1 || mo > 12 || d < 1 || d > 31) return null;
    const dt = new Date(y, mo - 1, d);
    if (dt.getFullYear() !== y || dt.getMonth() !== mo - 1 || dt.getDate() !== d) return null;
    return dt;
  }
  const fallback = new Date(raw);
  if (Number.isNaN(fallback.getTime())) return null;
  return new Date(fallback.getFullYear(), fallback.getMonth(), fallback.getDate());
}

export function ageFromBirthDate(birthDate: string | null | undefined, today = new Date()): number | null {
  const d = parseLocalDateInput(birthDate);
  if (!d) return null;
  let age = today.getFullYear() - d.getFullYear();
  const m = today.getMonth() - d.getMonth();
  if (m < 0 || (m === 0 && today.getDate() < d.getDate())) age -= 1;
  if (age < 10 || age > 100) return null;
  return age;
}

/** Birth year from YYYY-MM-DD (local parse). */
export function birthYearFromDate(birthDate: string | null | undefined): number | null {
  const d = parseLocalDateInput(birthDate);
  if (!d) return null;
  const y = d.getFullYear();
  return y >= 1900 && y <= 2100 ? y : null;
}

export function resolveAge(input: EnergyInput): number | null {
  const fromBirth = ageFromBirthDate(input.birthDate);
  if (fromBirth != null) return fromBirth;
  const age = Number(input.age);
  if (Number.isFinite(age) && age >= 10 && age <= 100) return age;
  return null;
}

export function resolveAdjustmentPct(input: EnergyInput): number {
  if (input.calorieAdjustmentPct != null && Number.isFinite(Number(input.calorieAdjustmentPct))) {
    return Math.max(-40, Math.min(40, Number(input.calorieAdjustmentPct)));
  }
  return GOAL_PCT[input.primaryGoal || "maintain"] ?? 0;
}

export function resolveActivity(input: EnergyInput): string {
  const level = (input.activityLevel || "").toLowerCase();
  if (level && ACTIVITY[level]) return level;
  const days = Number(input.daysPerWeek) || 3;
  if (days <= 2) return "light";
  if (days <= 4) return "moderate";
  if (days <= 5) return "active";
  return "very_active";
}

/** Normalize sex labels (en/ru) for Mifflin–St Jeor and calorie floor. */
export function isFemaleSex(sex?: string | null): boolean {
  const s = String(sex || "")
    .trim()
    .toLowerCase()
    .replaceAll("ё", "е");
  if (!s) return false;
  if (["f", "female", "woman", "w", "ж", "жен", "женский", "female_sex"].includes(s)) {
    return true;
  }
  if (s.startsWith("fem")) return true;
  if (s === "f") return true;
  if (s.startsWith("ж")) return true;
  return false;
}

/**
 * Live BMR/TDEE/target preview (mirrors backend).
 *
 * Mifflin–St Jeor (1990):
 *   Men:   10·W + 6.25·H − 5·A + 5
 *   Women: 10·W + 6.25·H − 5·A − 161
 * TDEE = BMR × activity factor (Harris–Benedict style multipliers).
 * Target = TDEE × (1 + adjustment%/100), floored at 1200♀ / 1500♂ kcal.
 */
export function previewEnergyTargets(input: EnergyInput) {
  const weight = Number(input.weightKg);
  const height = Number(input.heightCm);
  const age = resolveAge(input);
  if (!(weight > 0) || !(height > 0) || age == null) {
    return { complete: false as const, reason: "incomplete" as const };
  }
  const female = isFemaleSex(input.sex);
  const bmr = female
    ? 10 * weight + 6.25 * height - 5 * age - 161
    : 10 * weight + 6.25 * height - 5 * age + 5;
  const activity = resolveActivity(input);
  const tdee = bmr * ACTIVITY[activity];
  const adj = resolveAdjustmentPct(input);
  let target = tdee * (1 + adj / 100);
  target = Math.max(female ? 1200 : 1500, target);
  const proteinPerKg = input.primaryGoal === "lose_fat" ? 2 : input.primaryGoal === "gain_muscle" ? 1.8 : 1.6;
  const proteins = Math.max(80, proteinPerKg * weight);
  const fatRatio = input.primaryGoal === "gain_muscle" ? 0.25 : 0.3;
  const fats = Math.max(40, (target * fatRatio) / 9);
  const carbs = Math.max(0, (target - proteins * 4 - fats * 9) / 4);
  return {
    complete: true as const,
    bmr: Math.round(bmr),
    tdee: Math.round(tdee),
    caloriesTarget: Math.round(target),
    adjustmentPct: adj,
    activity,
    macros: {
      proteins: Math.round(proteins),
      fats: Math.round(fats),
      carbs: Math.round(carbs),
    },
  };
}

export const BODY_MEASURE_FIELDS = [
  { key: "weight_kg", label: "Вес, кг", unit: "кг", min: 20, max: 500 },
  { key: "neck_cm", label: "Шея, см", unit: "см", min: 1, max: 500 },
  { key: "shoulders_cm", label: "Плечи, см", unit: "см", min: 1, max: 500 },
  { key: "chest_cm", label: "Грудь, см", unit: "см", min: 1, max: 500 },
  { key: "waist_cm", label: "Талия, см", unit: "см", min: 1, max: 500 },
  { key: "hips_cm", label: "Бёдра, см", unit: "см", min: 1, max: 500 },
  { key: "bicep_cm", label: "Бицепс, см", unit: "см", min: 1, max: 500 },
  { key: "thigh_cm", label: "Бедро, см", unit: "см", min: 1, max: 500 },
  { key: "calf_cm", label: "Голень, см", unit: "см", min: 1, max: 500 },
] as const;

export const ACTIVITY_OPTIONS = [
  { id: "sedentary", label: "Сидячий" },
  { id: "light", label: "Лёгкая активность" },
  { id: "moderate", label: "Умеренная" },
  { id: "active", label: "Высокая" },
  { id: "very_active", label: "Очень высокая" },
] as const;
